src/morphological_operations.jl

```diff
@@ -99,3 +99,85 @@ which is defined as the arithmetic difference between the internal and the exter
 `region` allows you to control the dimensions over which this operation is performed.
 """
 morpholaplace(img::AbstractArray, region=coords_spatial(img)) = dilate(img, region) + erode(img, region) - 2img
+
+abstract type ThinAlgo end
+
+doc"""
+    struct GuoAlgo <: ThinAlgo end
+
+A thinning operation algorithm.
+
+The procedure carries on until there are no changes between two consecutives iterations.
+
+The algorithm is described in the the page 361 (Algorithm A1) of:
+* Guo, Z., & Hall, R. W. (1989). Parallel thinning with two-subiteration algorithms. Communications of the ACM, 32(3), 359-373.
+"""
+struct GuoAlgo <: ThinAlgo end
+
+"""
+```
+function thinning_iteration(img::AbstractArray{Bool,2}; odd_iteration::Bool, algo::GuoAlgo)
+```
+The thining iteration evaluates three conditions in order to determine which pixels of the image should be removed from img.
+The three conditions are explained in the page 361 of:
+* Guo, Z., & Hall, R. W. (1989). Parallel thinning with two-subiteration algorithms. Communications of the ACM, 32(3), 359-373.
+"""
+function thinning_iteration!(img_ori::AbstractArray{Bool,2}, odd_iteration::Bool, algo::GuoAlgo)
+    marker = trues(size(img_ori))
+    img = falses(size(img_ori).+2)
+    img[2:end-1,2:end-1] = img_ori
+    h, w = size(img)
+    for j=2:w-1, i=2:h-1
+        if !img[i,j]
+            continue
+        end
+        p1 = img[i-1,j-1]
+        p2 = img[i-1,j]
+        p3 = img[i-1,j+1]
+        p4 = img[i,j+1]
+        p5 = img[i+1,j+1]
+        p6 = img[i+1,j]
+        p7 = img[i+1,j-1]
+        p8 = img[i,j-1]
+        A = (!p2 && (p3 || p4)) + (!p4 && (p5 || p6)) + (!p6 && (p7 || p8)) + (!p8 && (p1 || p2))
+        B1 = (p1 || p2) + (p3 || p4) + (p5 || p6) + (p7 || p8)
```

Comment:
Consider renaming this variables to their names in the paper `N1`, `N2`, and `N`. Any other variable that you can map directly to the name in the paper will help other people reading and contributing to the code in the future.